include/ArrayVector.h

```diff
@@ -0,0 +1,388 @@
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#ifndef ARRAY_VECTOR_H
+#define ARRAY_VECTOR_H
+
+#include <algorithm>
+#include <cassert>
+#include <cstddef>
+#include <iterator>
+#include <memory>
+#include <new>
+#include <stdexcept>
+#include <utility>
+#include <type_traits>
+
+namespace lmms {
+
+namespace detail {
+
+template<typename T, typename = void>
+constexpr bool is_input_iterator_v = false;
+
+template<typename T>
+constexpr bool is_input_iterator_v<T, std::void_t<typename std::iterator_traits<T>::iterator_category>> =
+	std::is_convertible_v<typename std::iterator_traits<T>::iterator_category, std::input_iterator_tag>;
+
+} // namespace detail
+
+/**
+ * A container that stores up to a maximum of `N` elements of type `T` directly
+ * within itself, rather than separately on the heap. Useful when a dynamically
+ * resizeable container is needed for use in real-time code. Can be thought of
+ * as a hybrid between `std::array` and `std::vector`. The interface follows
+ * that of `std::vector` - see standard C++ documentation.
+ */
+template<typename T, std::size_t N>
+class ArrayVector
+{
+public:
+	using size_type = std::size_t;
+	using difference_type = std::ptrdiff_t;
+	using value_type = T;
+	using reference = T&;
+	using const_reference = const T&;
+	using pointer = T*;
+	using const_pointer = const T*;
+	using iterator = pointer;
+	using const_iterator = const_pointer;
+	using reverse_iterator = std::reverse_iterator<iterator>;
+	using const_reverse_iterator = std::reverse_iterator<const_iterator>;
+
+	ArrayVector() = default;
+
+	ArrayVector(const ArrayVector& other) noexcept(std::is_nothrow_copy_constructible_v<T>) :
+		m_size{other.m_size}
+	{
+		std::uninitialized_copy(other.begin(), other.end(), begin());
+	}
+
+	ArrayVector(ArrayVector&& other) noexcept(std::is_nothrow_move_constructible_v<T>) :
+		m_size{other.m_size}
+	{
+		std::uninitialized_move(other.begin(), other.end(), begin());
+		other.clear();
+	}
+
+	ArrayVector(size_type count, const T& value) noexcept(std::is_nothrow_copy_constructible_v<T>) :
+		m_size{count}
+	{
+		assert(count <= N);
+		std::uninitialized_fill_n(begin(), count, value);
+	}
+
+	explicit ArrayVector(size_type count) noexcept(std::is_nothrow_default_constructible_v<T>) :
+		m_size{count}
+	{
+		assert(count <= N);
+		std::uninitialized_value_construct_n(begin(), count);
```

Comment:
Wouldn't ``std::uninitialized_value_construct_n`` still allocate using ``new`` but just voidify the memory?

Reply:
It uses placement new, so there is no additional allocation involved.